File: app/src/main/java/me/zhanghai/android/files/provider/common/ForeignCopyMove.java
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.common;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

import androidx.annotation.NonNull;
import java8.nio.file.AtomicMoveNotSupportedException;
import java8.nio.file.CopyOption;
import java8.nio.file.FileAlreadyExistsException;
import java8.nio.file.Files;
import java8.nio.file.LinkOption;
import java8.nio.file.OpenOption;
import java8.nio.file.Path;
import java8.nio.file.StandardCopyOption;
import java8.nio.file.StandardOpenOption;
import java8.nio.file.attribute.BasicFileAttributeView;
import java8.nio.file.attribute.BasicFileAttributes;

class ForeignCopyMove {

    private ForeignCopyMove() {}

    public static void copy(@NonNull Path source, @NonNull Path target,
                            @NonNull CopyOption... options) throws IOException {
        CopyOptions copyOptions = CopyOptions.fromArray(options);
        if (copyOptions.hasAtomicMove()) {
            throw new UnsupportedOperationException(StandardCopyOption.ATOMIC_MOVE.toString());
        }
        LinkOption[] linkOptions = copyOptions.hasNoFollowLinks() ?
                new LinkOption[] { LinkOption.NOFOLLOW_LINKS } : new LinkOption[0];
        BasicFileAttributes sourceAttributes = Files.readAttributes(source,
                BasicFileAttributes.class, linkOptions);
        if (!(sourceAttributes.isRegularFile() || sourceAttributes.isDirectory()
                || sourceAttributes.isSymbolicLink())) {
            throw new IOException("Cannot copy special file to foreign provider");
        }
        if (!copyOptions.hasReplaceExisting() && Files.exists(target)) {
            throw new FileAlreadyExistsException(source.toString(), target.toString(), null);
        }
        if (sourceAttributes.isRegularFile()) {
            if (copyOptions.hasReplaceExisting()) {
                Files.deleteIfExists(target);
            }
            OpenOption[] openOptions = copyOptions.hasNoFollowLinks() ?
                    new OpenOption[] { LinkOption.NOFOLLOW_LINKS } : new OpenOption[0];
            try (InputStream inputStream = MoreFiles.newInputStream(source, openOptions)) {
                OutputStream outputStream = MoreFiles.newOutputStream(target,
                        StandardOpenOption.CREATE_NEW, StandardOpenOption.WRITE);
                boolean successful = false;
                try {
                    MoreFiles.copy(inputStream, outputStream, copyOptions.getProgressListener(),
                            copyOptions.getProgressIntervalMillis());
                    successful = true;
                } finally {
                    try {
                        outputStream.close();
                    } finally {
                        if (!successful) {
                            try {
                                Files.deleteIfExists(target);
                            } catch (IOException | UnsupportedOperationException e) {
                                e.printStackTrace();
                            }
                        }
                    }
                }
            }
        } else if (sourceAttributes.isDirectory()) {
            if (copyOptions.hasReplaceExisting()) {
                Files.deleteIfExists(target);
            }
            Files.createDirectory(target);
            if (copyOptions.hasProgressListener()) {
                copyOptions.getProgressListener().accept(sourceAttributes.size());
            }
        } else if (sourceAttributes.isSymbolicLink()) {
            Path sourceTarget = Files.readSymbolicLink(source);
            try {
                // Might throw UnsupportedOperationException, so we cannot delete beforehand.
                Files.createSymbolicLink(target, sourceTarget);
            } catch (FileAlreadyExistsException e) {
                if (!copyOptions.hasReplaceExisting()) {
                    throw e;
                }
                Files.deleteIfExists(target);
                Files.createSymbolicLink(target, sourceTarget);
            }
            if (copyOptions.hasProgressListener()) {
                copyOptions.getProgressListener().accept(sourceAttributes.size());
            }
        } else {
            throw new AssertionError();
        }
        if (copyOptions.hasCopyAttributes()) {
            // We don't take error when copying attribute fatal, so errors will only be logged from
            // now on.
            BasicFileAttributeView targetAttributeView = Files.getFileAttributeView(target,
                    BasicFileAttributeView.class);
            try {
                targetAttributeView.setTimes(sourceAttributes.lastModifiedTime(),
                        sourceAttributes.lastAccessTime(), sourceAttributes.creationTime());
            } catch (IOException | UnsupportedOperationException e) {
                e.printStackTrace();
            }
        }
    }

    public static void move(@NonNull Path source, @NonNull Path target,
                            @NonNull CopyOption... options) throws IOException {
        CopyOptions copyOptions = CopyOptions.fromArray(options);
        if (copyOptions.hasAtomicMove()) {
            throw new AtomicMoveNotSupportedException(source.toString(), target.toString(),
                    "Cannot move file atomically to foreign provider");
        }
        CopyOption[] optionsForCopy;
        if (copyOptions.hasCopyAttributes() && copyOptions.hasNoFollowLinks()) {
            optionsForCopy = options;
        } else {
            optionsForCopy = new CopyOptions(copyOptions.hasReplaceExisting(), true, false, true,
                    copyOptions.getProgressListener(), copyOptions.getProgressIntervalMillis())
                    .toArray();
        }
        copy(source, target, optionsForCopy);
        try {
            Files.delete(source);
        } catch (IOException | UnsupportedOperationException e) {
            try {
                Files.delete(target);
            } catch (IOException | UnsupportedOperationException e2) {
                e.addSuppressed(e2);
            }
            throw e;
        }
    }
}

File: app/src/main/java/me/zhanghai/android/files/provider/common/ForeignCopyMove.kt
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.common

import java8.nio.file.AtomicMoveNotSupportedException
import java8.nio.file.CopyOption
import java8.nio.file.FileAlreadyExistsException
import java8.nio.file.LinkOption
import java8.nio.file.NoSuchFileException
import java8.nio.file.Path
import java8.nio.file.StandardCopyOption
import java8.nio.file.StandardOpenOption
import java8.nio.file.attribute.BasicFileAttributeView
import java8.nio.file.attribute.BasicFileAttributes
import java.io.IOException

internal object ForeignCopyMove {
    @Throws(IOException::class)
    fun copy(source: Path, target: Path, vararg options: CopyOption) {
        val copyOptions = options.toCopyOptions()
        if (copyOptions.atomicMove) {
            throw UnsupportedOperationException(StandardCopyOption.ATOMIC_MOVE.toString())
        }
        val linkOptions = if (copyOptions.noFollowLinks) {
            arrayOf(LinkOption.NOFOLLOW_LINKS)
        } else {
            emptyArray()
        }
        val sourceAttributes = source.readAttributes(BasicFileAttributes::class.java, *linkOptions)
        if (!(sourceAttributes.isRegularFile || sourceAttributes.isDirectory
                || sourceAttributes.isSymbolicLink)) {
            throw IOException("Cannot copy special file to foreign provider")
        }
        if (!copyOptions.replaceExisting && target.exists(LinkOption.NOFOLLOW_LINKS)) {
            throw FileAlreadyExistsException(source.toString(), target.toString(), null)
        }
        when {
            sourceAttributes.isRegularFile -> {
                if (copyOptions.replaceExisting) {
                    target.deleteIfExists()
                }
                val openOptions = if (copyOptions.noFollowLinks) {
                    arrayOf(LinkOption.NOFOLLOW_LINKS)
                } else {
                    emptyArray()
                }
                source.newInputStream(*openOptions).use { inputStream ->
                    val outputStream = target.newOutputStream(
                        StandardOpenOption.CREATE_NEW, StandardOpenOption.WRITE
                    )
                    var successful = false
                    try {
                        inputStream.copyTo(
                            outputStream, copyOptions.progressIntervalMillis,
                            copyOptions.progressListener
                        )
                        successful = true
                    } finally {
                        try {
                            outputStream.close()
                        } finally {
                            if (!successful) {
                                try {
                                    target.deleteIfExists()
                                } catch (e: IOException) {
                                    e.printStackTrace()
                                } catch (e: UnsupportedOperationException) {
                                    e.printStackTrace()
                                }
                            }
                        }
                    }
                }
            }
            sourceAttributes.isDirectory -> {
                if (copyOptions.replaceExisting) {
                    target.deleteIfExists()
                }
                target.createDirectory()
                copyOptions.progressListener?.invoke(sourceAttributes.size())
            }
            sourceAttributes.isSymbolicLink -> {
                val sourceTarget = source.readSymbolicLink()
                try {
                    // Might throw UnsupportedOperationException, so we cannot delete beforehand.
                    target.createSymbolicLink(sourceTarget)
                } catch (e: FileAlreadyExistsException) {
                    if (!copyOptions.replaceExisting) {
                        throw e
                    }
                    target.deleteIfExists()
                    target.createSymbolicLink(sourceTarget)
                }
                copyOptions.progressListener?.invoke(sourceAttributes.size())
            }
            else -> throw AssertionError()
        }
        // We don't take error when copying attribute fatal, so errors will only be logged from
        // now on.
        val targetAttributeView = target.getFileAttributeView(BasicFileAttributeView::class.java)!!
        val lastModifiedTime = sourceAttributes.lastModifiedTime()
        val lastAccessTime = if (copyOptions.copyAttributes) {
            sourceAttributes.lastAccessTime()
        } else {
            null
        }
        val creationTime = if (copyOptions.copyAttributes) sourceAttributes.creationTime() else null
        try {
            targetAttributeView.setTimes(lastModifiedTime, lastAccessTime, creationTime)
        } catch (e: IOException) {
            e.printStackTrace()
        } catch (e: UnsupportedOperationException) {
            e.printStackTrace()
        }
    }

    @Throws(IOException::class)
    fun move(source: Path, target: Path, vararg options: CopyOption) {
        val copyOptions = options.toCopyOptions()
        if (copyOptions.atomicMove) {
            throw AtomicMoveNotSupportedException(
                source.toString(), target.toString(),
                "Cannot move file atomically to foreign provider"
            )
        }
        val optionsForCopy = if (copyOptions.copyAttributes && copyOptions.noFollowLinks) {
            options
        } else {
            CopyOptions(
                copyOptions.replaceExisting, true, false, true, copyOptions.progressIntervalMillis,
                copyOptions.progressListener
            ).toArray()
        }
        copy(source, target, *optionsForCopy)
        try {
            source.delete()
        } catch (e: IOException) {
            if (e !is NoSuchFileException) {
                try {
                    target.delete()
                } catch (e2: IOException) {
                    e.addSuppressed(e2)
                } catch (e2: UnsupportedOperationException) {
                    e.addSuppressed(e2)
                }
            }
            throw e
        } catch (e: UnsupportedOperationException) {
            try {
                target.delete()
            } catch (e2: IOException) {
                e.addSuppressed(e2)
            } catch (e2: UnsupportedOperationException) {
                e.addSuppressed(e2)
            }
            throw e
        }
    }
}
